File: java-automation/src/test/java/com/lineasupply/automation/stepdefinitions/FilterSteps.java
package com.lineasupply.automation.stepdefinitions;

import com.lineasupply.automation.driver.DriverFactory;
import com.lineasupply.automation.dsl.FilterDsl;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class FilterSteps {

    private final FilterDsl filter = new FilterDsl(DriverFactory.create());

    @Given("the shopper is on the homepage with products visible")
    public void shopperOnHomepageWithProductsVisible() {
        filter.openHomePageAndWaitForProducts();
    }

    @When("the shopper searches for the first product name")
    public void shopperSearchesForFirstProductName() {
        filter.searchForFirstProductName();
    }

    @When("the shopper searches for {string}")
    public void shopperSearchesFor(String term) {
        filter.searchFor(term);
    }

    @Then("the URL should contain the search term")
    public void urlShouldContainSearchTerm() {
        filter.assertUrlContainsCapturedTerm();
    }

    @Then("search results should be displayed")
    public void searchResultsShouldBeDisplayed() {
        filter.assertSearchResultsDisplayed();
    }

    @Then("no results or empty state should be shown")
    public void noResultsOrEmptyStateShouldBeShown() {
        filter.assertEmptyOrNoResults();
    }
}

File: typescript-automation/src/steps/filterSteps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import { expect } from "chai";
import type { MyWorld } from "../support/world";

let capturedSearchTerm = "";

Given<MyWorld>("the shopper is on the homepage with products visible", async function () {
  await this.catalogue.browseCatalogue();
});

When<MyWorld>("the shopper searches for the first product name", async function () {
  const { cards } = await this.catalogue.getProductListing();
  expect(cards, "product cards must be visible to search").to.not.be.empty;
  capturedSearchTerm = (cards[0]?.title ?? "").split(" ")[0] ?? "";
  await this.catalogue.searchFor(capturedSearchTerm);
});

When<MyWorld>("the shopper searches for {string}", async function (term: string) {
  capturedSearchTerm = term;
  await this.catalogue.searchFor(term);
});

Then<MyWorld>("the URL should contain the search term", async function () {
  expect(await this.catalogue.currentUrl(), "URL should contain the searched term").to.include(
    `/search/${capturedSearchTerm}`
  );
});

Then<MyWorld>("search results should be displayed", async function () {
  expect((await this.catalogue.getSearchResults()).cards, "search results should show at least one card").to.not.be
    .empty;
});

Then<MyWorld>("no results or empty state should be shown", async function () {
  const results = await this.catalogue.getSearchResults();
  expect(results.emptyStateVisible || results.cards.length === 0, "either no-results or zero cards expected").to.be
    .true;
});
